File: api/core/hunt/ingestion/tools/whois.py
"""WHOIS / RDAP lookup.

Tries RDAP first (returns structured JSON with registrant
information), then falls back to a text-based WHOIS server.

No API key required. Works for domains.
"""
from __future__ import annotations

import json
import re
from typing import Any, List, Mapping

import httpx

from hunt.ingestion.base import Finding, ToolFunction
from hunt.models import EntityKind, SourceTool


_EMAIL_RE = re.compile(r"[\w.+-]+@[\w-]+(?:\.[\w-]+)+")


class WhoisTool(ToolFunction):
    name = SourceTool.WHOIS
    accepts = {"domain"}
    emits = {EntityKind.ORG, EntityKind.EMAIL, EntityKind.PERSON, EntityKind.DOMAIN}
    description = "RDAP/WHOIS registrant + registrar + email extraction."
    per_request_timeout = 10.0

    RDAP_URL = "https://rdap.org/domain/{target}"
    WHOIS_FALLBACK_HOST = "whois.iana.org"

    def _fetch(self, query: Mapping[str, Any]) -> httpx.Response:
        target = query["target"].strip().lower()
        with httpx.Client(timeout=self.per_request_timeout) as client:
            try:
                resp = client.get(self.RDAP_URL.format(target=target))
                if resp.status_code < 400:
                    resp.headers["content-type"] = (
                        resp.headers.get("content-type") or "application/rdap+json"
                    )
                    return resp
            except httpx.HTTPError:
                pass
            # Fallback: text WHOIS over the IANA referral server.
            with client.stream(
                "GET", f"https://{self.WHOIS_FALLBACK_HOST}/{target}"
            ) as r:
                body = r.read()
            return httpx.Response(
                200, content=body, headers={"content-type": "text/plain"}
            )

    def _parse(self, raw: bytes, query: Mapping[str, Any]) -> List[Finding]:
        target = query["target"].strip().lower()
        findings: List[Finding] = []
        try:
            data = json.loads(raw.decode("utf-8"))
        except (UnicodeDecodeError, json.JSONDecodeError):
            data = None

        if isinstance(data, dict) and "entities" in data:
            for entity in data.get("entities", []):
                roles = entity.get("roles", [])
                vcard = entity.get("vcardArray", [None, []])
                name = ""
                email = ""
                if isinstance(vcard, list) and len(vcard) > 1:
                    for entry in vcard[1]:
                        if entry and entry[0] == "fn":
                            name = entry[3]
                        if entry and entry[0] == "email":
                            email = entry[3]
                if "registrant" in roles and email:
                    findings.append(
                        Finding(
                            source_tool=self.name,
                            entity_kind=EntityKind.EMAIL,
                            entity_value=email,
                            attributes={
                                "role": "registrant",
                                "name": name,
                                "domain": target,
                            },
                        )
                    )
                if "registrar" in roles and name:
                    findings.append(
                        Finding(
                            source_tool=self.name,
                            entity_kind=EntityKind.ORG,
                            entity_value=name,
                            attributes={"role": "registrar", "domain": target},
                        )
                    )
            return findings

        text = raw.decode("utf-8", errors="replace")
        for m in _EMAIL_RE.findall(text):
            findings.append(
                Finding(
                    source_tool=self.name,
                    entity_kind=EntityKind.EMAIL,
                    entity_value=m.lower(),
                    attributes={"role": "registrant", "domain": target, "source": "whois-text"},
                )
            )
        m = re.search(r"Registrar:\s*(.+)", text)
        if m:
            findings.append(
                Finding(
                    source_tool=self.name,
                    entity_kind=EntityKind.ORG,
                    entity_value=m.group(1).strip(),
                    attributes={"role": "registrar", "domain": target},
                )
            )
        return findings

File: lib/tools/whois.ts
/**
 * WHOIS / RDAP lookup.
 *
 * Tries RDAP first (returns structured JSON with registrant info),
 * then falls back to a text-based WHOIS server.
 *
 * No API key required. Works for domains.
 */

import { ToolFunction, type FetchResult, type Finding } from "./base";

const EMAIL_RE = /[\w.+-]+@[\w-]+(?:\.[\w-]+)+/;

export class WhoisTool extends ToolFunction {
  readonly name = "whois";
  readonly accepts = new Set(["domain"]);
  readonly emits = new Set(["org", "email", "person", "domain"]);
  readonly description = "RDAP/WHOIS registrant + registrar + email extraction.";
  readonly per_request_timeout_ms = 10_000;

  private static RDAP_URL = "https://rdap.org/domain/{target}";
  private static WHOIS_HOST = "https://whois.iana.org/{target}";

  protected async _fetch(query: { target: string }): Promise<FetchResult> {
    const target = (query.target || "").trim().toLowerCase();
    // Try RDAP first.
    try {
      const r = await fetch(
        WhoisTool.RDAP_URL.replace("{target}", target),
        { signal: AbortSignal.timeout(this.per_request_timeout_ms) }
      );
      if (r.ok) {
        const bytes = Buffer.from(await r.arrayBuffer());
        return {
          bytes,
          contentType: r.headers.get("content-type") || "application/rdap+json",
        };
      }
    } catch {
      // fall through to text WHOIS
    }
    const r2 = await fetch(WhoisTool.WHOIS_HOST.replace("{target}", target), {
      signal: AbortSignal.timeout(this.per_request_timeout_ms),
    });
    return { bytes: Buffer.from(await r2.arrayBuffer()), contentType: "text/plain" };
  }

  protected _parse(raw: Buffer, query: { target: string }): Finding[] {
    const target = (query.target || "").trim().toLowerCase();
    const findings: Finding[] = [];
    let data: any = null;
    try {
      data = JSON.parse(raw.toString("utf-8"));
    } catch {
      data = null;
    }

    if (data && Array.isArray(data.entities)) {
      for (const entity of data.entities as any[]) {
        const roles: string[] = entity.roles || [];
        const vcard = entity.vcardArray || [null, []];
        let name = "";
        let email = "";
        if (Array.isArray(vcard) && Array.isArray(vcard[1])) {
          for (const entry of vcard[1] as any[][]) {
            if (entry && entry[0] === "fn") name = String(entry[3] ?? "");
            if (entry && entry[0] === "email") email = String(entry[3] ?? "");
          }
        }
        if (roles.includes("registrant") && email) {
          findings.push({
            source_tool: this.name,
            entity_kind: "email",
            entity_value: email,
            attributes: { role: "registrant", name, domain: target },
          });
        }
        if (roles.includes("registrar") && name) {
          findings.push({
            source_tool: this.name,
            entity_kind: "org",
            entity_value: name,
            attributes: { role: "registrar", domain: target },
          });
        }
      }
      return findings;
    }

    const text = raw.toString("utf-8");
    for (const m of text.matchAll(EMAIL_RE)) {
      findings.push({
        source_tool: this.name,
        entity_kind: "email",
        entity_value: m[0].toLowerCase(),
        attributes: { role: "registrant", domain: target, source: "whois-text" },
      });
    }
    const registrar = /Registrar:\s*(.+)/.exec(text);
    if (registrar) {
      findings.push({
        source_tool: this.name,
        entity_kind: "org",
        entity_value: registrar[1].trim(),
        attributes: { role: "registrar", domain: target },
      });
    }
    return findings;
  }
}
